lib/Interpreter/CIFactory.cpp

```diff
@@ -338,9 +339,367 @@ static bool getVisualStudioDir(std::string &path) {
+
+  // Try to get the SDK for whatever version of OS X is currently running
+  if (void *core = dlopen(
+          "/System/Library/Frameworks/CoreServices.framework/CoreServices",
+          RTLD_LAZY)) {
+    SInt32 majorVersion = -1, minorVersion = -1;
+    typedef ::OSErr (*GestaltProc)(::OSType, ::SInt32 *);
+    if (GestaltProc Gestalt = (GestaltProc)dlsym(core, "Gestalt")) {
+      Gestalt(gestaltSystemVersionMajor, &majorVersion);
+      Gestalt(gestaltSystemVersionMinor, &minorVersion);
+    }
+    ::dlclose(core);
+
+    if (majorVersion != -1 && minorVersion != -1) {
+      if (getISysRootVersion(SDKs, majorVersion, minorVersion, sysRoot))
+        return true;
+    }
+  }
+
+#define GET_ISYSROOT_VER(maj, min) \
+  if (getISysRootVersion(SDKs, maj, min, sysRoot)) \
+    return true;
+
+  // Try to get the SDK for whatever cling was compiled with
+  #if defined(MAC_OS_X_VERSION_10_11)
+    GET_ISYSROOT_VER(10, 11);
+  #elif defined(MAC_OS_X_VERSION_10_10)
+    GET_ISYSROOT_VER(10, 10);
+  #elif defined(MAC_OS_X_VERSION_10_9)
+    GET_ISYSROOT_VER(10, 9);
+  #elif defined(MAC_OS_X_VERSION_10_8)
+    GET_ISYSROOT_VER(10, 8);
+  #elif defined(MAC_OS_X_VERSION_10_7)
+    GET_ISYSROOT_VER(10, 7);
+  #elif defined(MAC_OS_X_VERSION_10_6)
+    GET_ISYSROOT_VER(10, 6);
+  #elif defined(MAC_OS_X_VERSION_10_5)
+    GET_ISYSROOT_VER(10, 5);
+  #elif defined(MAC_OS_X_VERSION_10_4)
+    GET_ISYSROOT_VER(10, 4);
+  #elif defined(MAC_OS_X_VERSION_10_3)
+    GET_ISYSROOT_VER(10, 3);
+  #elif defined(MAC_OS_X_VERSION_10_2)
+    GET_ISYSROOT_VER(10, 2);
+  #elif defined(MAC_OS_X_VERSION_10_1)
+    GET_ISYSROOT_VER(10, 1);
+  #else // MAC_OS_X_VERSION_10_0
+    GET_ISYSROOT_VER(10, 0);
+  #endif
+
+#undef GET_ISYSROOT_VER
+
+  // Scan the SDKs directory and use the latest
+  std::error_code ec;
+  std::vector<std::string> srtd;
+  for (fs::directory_iterator it(SDKs, ec), e; it != e; it.increment(ec))
+    srtd.push_back(it->path());
+  if (!srtd.empty()) {
+    std::sort(srtd.begin(), srtd.end(), std::greater<std::string>());
+    sysRoot.swap(srtd[0]);
+    return true;
+  }
+
+  return false;
+}
+
+#endif // __APPLE__, _MSC_VER
+
+using namespace clang;
+using namespace cling;
 
 namespace {
+  // This function isn't referenced outside its translation unit, but it
+  // can't use the "static" keyword because its address is used for
+  // GetMainExecutable (since some platforms don't support taking the
+  // address of main, and some platforms can't implement GetMainExecutable
+  // without being given the address of a function in the main executable).
+  std::string GetExecutablePath(const char *Argv0) {
+    // This just needs to be some symbol in the binary; C++ doesn't
+    // allow taking the address of ::main however.
+    void *MainAddr = (void*) intptr_t(GetExecutablePath);
+    return llvm::sys::fs::getMainExecutable(Argv0, MainAddr);
+  }
+
+  class AdditionalArgList {
+    typedef std::vector< std::pair<const char*,std::string> > container_t;
```

Comment:
I left a comment here but it seems gone now... Could you rename this to something like `AdditionalArgList`.